### ISSUE TITLE
lowering for `oklch` does not work when using numeric or mixed values

### ISSUE THREAD

**Issue by mayank99** [parcel-bundler/lightningcss#809]
Currently, lightningcss lowers the following correctly (albeit with an unnecessary conversion from `oklch` to `lab`):

```css
--color-1: oklch(100% none none);
--color-2: oklch(100% 0 0);
```

Whereas the following are ignored completely:

```css
--color-3: oklch(1 0 0);
--color-4: oklch(100% 0% 0deg);
```

See [playground](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Afalse%2C%22customMedia%22%3Afalse%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A7208960%2C%22firefox%22%3A7208960%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22body%20%7B%5Cn%20%20--color-1%3A%20oklch(100%25%20none%20none)%3B%5Cn%20%20--color-2%3A%20oklch(100%25%200%200)%3B%5Cn%20%20--color-3%3A%20oklch(1%200%200)%3B%5Cn%20%20--color-4%3A%20oklch(100%25%200%25%200deg)%3B%5Cn%7D%5Cn%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D).

I came across a similar issue (#445) which was closed by https://github.com/parcel-bundler/lightningcss/commit/835e5bc68da7ecc7c2e59b20617eed56465f9045, but it doesn't work in the latest version.

**Comment by mattpilott** [comment]
Id also like to understand to oklch to lab, was expecting to still see oklch in the final css 🤓

**Comment by devongovett** [comment]
That commit was reverted in 81cd955 . At some point the spec changed to accept numbers or percentages for all components instead of one or the other. Unfortunately this is a breaking change for Lightning CSS because it changes the way calcs work. There is a PR #465 to update to the latest spec, but it is now waiting for a major version release unfortunately. In the meantime, you can use percentages. 

As for the conversion to lab, that's because you have specified browser targets that do not support `oklch` but do support lab. According to MDN, `oklch` is only supported in Chrome 111 and Firefox 113, but the targets in the playground are set to 110 for both.

**Comment by mayank99** [comment]
Thanks for the response. That's unfortunate about the breaking change, but using percentages is not a big deal I suppose.

I'm not sure that the reasoning for the `lab` issue is correct. `lab` and `oklch` have equal browser support, afaict (see [caniuse](https://caniuse.com/css-lch-lab)). Even if I lower the targets some more, `lab` is still used (see [playground](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Afalse%2C%22customMedia%22%3Afalse%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6553600%2C%22firefox%22%3A6553600%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22body%20%7B%5Cn%20%20--color-1%3A%20oklch(100%25%20none%20none)%3B%5Cn%20%20--color-2%3A%20oklch(100%25%200%200)%3B%5Cn%20%20--color-3%3A%20oklch(1%200%200)%3B%5Cn%20%20--color-4%3A%20oklch(100%25%200%25%200deg)%3B%5Cn%7D%5Cn%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D)). The fallback for older browsers is not the `lab` but rather the hex colors. So the correct combination would be hex (fallback) + `oklch` (untransformed).

**Comment by devongovett** [comment]
That's true. I guess Safari was the difference. 15 supported lab and 15.4 added support for oklab. In this case you aren't targeting any Safari browsers so it could probably be avoided.

**Comment by nicksrandall** [comment]
Tailwind CSS v4 colors now use oklch colors by default and they are not being lowered because they use the decimal syntax and not the percentage syntax.

**Comment by Eptagone** [comment]
Hi, I'm using TailwindCSS v4 too and like @nicksrandall said, it doesn't work because TailwindCSS use decimal syntax but I also discovered that lightningcss is taking the oklch color as a function instead when it use decimals. So, I found a temporary solution by creating a custom visitor to transform the function into a color as it should be.

```typescript
import {
    type CustomAtRules, type Function, type ReturnedDeclaration, type TokenOrValue, type Visitor,
} from "lightningcss";

function transformFunctionIntoColor(tokenOrValue: TokenOrValue & { type: "function"; value: Function }): TokenOrValue {
    let [l, c, h, alpha] = tokenOrValue.value.arguments
        .filter((arg): arg is TokenOrValue & { value: { type: "number"; value: number } } => arg.type === "token" && arg.value.type === "number")
        .map(arg => arg.value.value);
    l ??= 0;
    c ??= 0;
    h ??= 0;
    alpha ??= 1;

    const oklchColor: TokenOrValue = {
        type: "color",
        value: {
            type: "oklch",
            l, c, h, alpha,
        },
    };
    return oklchColor;
}

/**
 * Fix oklch colors which are detected as functions instead of colors.
 */
const FixOklchColorsVisitor: Visitor<CustomAtRules> = {
    Declaration(declaration): ReturnedDeclaration | ReturnedDeclaration[] | void {
        let needsUpdate = false;
        if (declaration.property === "custom") {
            for (let index = 0; index < declaration.value.value.length; index++) {
                const tokenOrValue = declaration.value.value[index];
                if (tokenOrValue?.type === "function" && tokenOrValue.value.name === "oklch") {
                    declaration.value.value[index] = transformFunctionIntoColor(tokenOrValue);
                    needsUpdate = true;
                }
            }
        }
        if (needsUpdate) {
            return declaration;
        }
    },
};
```

Now, the visitor have to be added in the lightningcss options along other visitors. In my case, I'm using Astro, so I added into the vite css options.
```typescript
export default defineConfig({
    // ...
    vite: {
        css: {
            transformer: "lightningcss",
            lightningcss: {
                // ...
                visitor: composeVisitors([
                    // other visitors
                    FixOklchColorsVisitor, // Here with the others visitors
                ]),
            },
        },
        plugins: [tailwindcss()],
    },
});
```
